p2p/transport/quic/virtuallistener.go

```diff
@@ -0,0 +1,168 @@
+package libp2pquic
+
+import (
+	"errors"
+	"sync"
+
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	"github.com/libp2p/go-libp2p/p2p/transport/quicreuse"
+	"github.com/lucas-clemente/quic-go"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const acceptBufferPerVersion = 4
+
+// virtualListener is a listener that exposes a single multiaddr but uses another listener under the hood
+type virtualListener struct {
+	*listener
+	udpAddr       string
+	version       quic.VersionNumber
+	t             *transport
+	acceptRunnner *acceptLoopRunner
+	acceptChan    chan acceptVal
+}
+
+var _ tpt.Listener = &virtualListener{}
+
+func (l *virtualListener) Multiaddr() ma.Multiaddr {
+	return l.listener.localMultiaddrs[l.version]
+}
+
+func (l *virtualListener) Close() error {
+	l.acceptRunnner.rmAcceptForVersion(l.version)
+
+	l.t.listenersMu.Lock()
+	defer l.t.listenersMu.Unlock()
+
+	var err error
+	listeners := l.t.listeners[l.udpAddr]
+	if len(listeners) == 1 {
+		// This is the last virtual listener here, so we can close the underlying listener
+		err = l.listener.Close()
+		delete(l.t.listeners, l.udpAddr)
+	} else {
+		for i := 0; i < len(listeners); i++ {
+			// Swap remove
+			if l == listeners[i] {
+				listeners[i] = listeners[len(listeners)-1]
+				listeners = listeners[0 : len(listeners)-1]
+				l.t.listeners[l.udpAddr] = listeners
+				break
+			}
+		}
+	}
+
+	return err
+}
+
+func (l *virtualListener) Accept() (tpt.CapableConn, error) {
+	var v acceptVal
+	var ok bool
+	select {
+	// Check if we have a pending connection first
+	case v, ok = <-l.acceptChan:
+	default:
+		// No? Let's call Accept and wait for a connection
+		go l.acceptRunnner.Accept(l.listener, l.version)
+		v, ok = <-l.acceptChan
+	}
+	if !ok {
+		return nil, errors.New("listener closed")
+	}
+
+	return v.conn, v.err
+}
+
+type acceptVal struct {
+	conn tpt.CapableConn
+	err  error
+}
+
+type acceptLoopRunner struct {
+	listenerMu sync.Mutex
```

Comment:
Why is this mutex needed? I assume it's to protect against concurrent calls to `Accept` from multiple listeners? If so, I'd prefer to call this a semaphore (and make it a `sem chan struct` with cap 1).